src/dune/file_tree.ml

```diff
@@ -306,76 +370,123 @@ module rec Memoized : sig
 end = struct
   open Memoized
 
-  let get_sub_dirs ~dirs_visited ~dirs ~sub_dirs
-      ~(dir_status : Sub_dirs.Status.t) =
-    let sub_dirs =
-      Sub_dirs.eval sub_dirs ~dirs:(List.map ~f:(fun (a, _, _) -> a) dirs)
-    in
-    dirs
-    |> List.fold_left ~init:(String.Map.empty, String.Map.empty)
-         ~f:(fun (dirs_visited_acc, subdirs) (fn, path, file) ->
-           let status = Sub_dirs.status sub_dirs ~dir:fn in
-           match status with
-           | Ignored -> (dirs_visited_acc, subdirs)
-           | Status status ->
-             let dir_status : Sub_dirs.Status.t =
-               match (dir_status, status) with
-               | Data_only, _ -> Data_only
-               | Vendored, Normal -> Vendored
-               | _, _ -> status
-             in
-             let dirs_visited_acc =
-               if Sys.win32 then
-                 dirs_visited_acc
-               else
-                 let new_dirs_visited =
-                   File.Map.update dirs_visited file ~f:(function
-                     | None -> Some path
-                     | Some first_path ->
-                       User_error.raise
-                         [ Pp.textf
-                             "Path %s has already been scanned. Cannot scan it \
-                              again through symlink %s"
-                             (Path.Source.to_string_maybe_quoted first_path)
-                             (Path.Source.to_string_maybe_quoted path)
-                         ])
-                 in
-                 String.Map.add_exn dirs_visited_acc fn new_dirs_visited
-             in
-             let sub_dir =
-               let sub_dir_as_t = find_dir_raw path in
-               { Dir0.sub_dir_status = dir_status; sub_dir_as_t }
-             in
-             let subdirs = String.Map.set subdirs fn sub_dir in
-             (dirs_visited_acc, subdirs))
+  module Get_subdir : sig
+    (** Get all the sub directories of [path].*)
+    val all :
+         dirs_visited:Dirs_visited.t
+      -> dirs:(string * Path.Source.t * File.t) list
+      -> sub_dirs:Predicate_lang.Glob.t Sub_dirs.Status.Map.t
+      -> parent_status:Sub_dirs.Status.t
+      -> dune_file:Dune_file.t option (** to interpret [(chdir ..)] stanzas *)
+      -> path:Path.Source.t
+      -> Dirs_visited.Per_fn.t * Dir0.sub_dir String.Map.t
+  end = struct
+    let status ~status_map ~(parent_status : Sub_dirs.Status.t) dir :
+        Sub_dirs.Status.t option =
+      let status = Sub_dirs.status status_map ~dir in
+      match status with
+      | Ignored -> None
+      | Status status ->
+        Some
+          ( match (parent_status, status) with
+          | Data_only, _ -> Data_only
+          | Vendored, Normal -> Vendored
+          | _, _ -> status )
```

Comment:
Might be worth spelling out the cases here and maybe leaving a comment on why the choices make sense.

By the way, if the directory is data-only then it shouldn't have any rules generated, right? In particular, its dune files are not interpreted and, presumably, the "chdir" in the parent should be forbidden. If that's the case, can `Data_only, _` case be visited at all?

Reply:
> By the way, if the directory is data-only then it shouldn't have any rules generated, right? In particular, its dune files are not interpreted and, presumably, the "chdir" in the parent should be forbidden. If that's the case, can Data_only, _ case be visited at all?

It would be possible to restrict this, but I have the following a use case where I ignore the dune file in a sub directory by marking it as `data_only_dirs`, and then inserting my own dune file in there. So I think that `data-only` should be interpreted as having no `dune` file in it, rather than not having any stanzas or  rules.

Reply:
Oh, I didn't realize we were allowing this. I guess that's fine too.